design-documents/graph-ql/nullability.md

```diff
@@ -0,0 +1,216 @@
+
+- A nullable parent field is found
+- We reach the root `Query` object
+
+ A non-nullable type in the wrong place can end up causing a client to lose more data farther up the response tree, even though that data may have been resolved without errors.
+
+_[Further reading on the relationship between non-null types and errors](http://spec.graphql.org/draft/#sec-Errors-and-Non-Nullability)_
+
+### Example: Bad use of non-nullable field
+
+In this example, a client wants to query for data to render a product details page. But, because the `related_products` field is non-nullable, the client will lose access to data like `name` and `price` if `related_products` throws any errors.
+
+Ideally, a UI would get the critical product data, render the page, and just not render the "Related Products" display.
+
+**Schema**:
+```graphql
+type Query {
+   product(id: ID): Product
+}
+
+type Product {
+    id: ID!
+    name: String!
+    price: Money!
+    description: String!
+    url: String!
+    related_products: RelatedProducts!
+}
+```
+**Query**:
+```graphql
+query ProductDetailsPage($id: ID) {
+    product(id: $id) {
+       id
+       name
+       price
+       description
+       # related_products can't be null. When the backing service is
+       # down/unreachable at the time this query runs, we can't fulfill
+       # the contract
+       related_products {
+           items {
+               id
+               name
+               price
+               url
+           }
+       }
+    }
+}
+```
+
+**Response**
+
+Note that no data was obtained by the client for `product`
+```js
+{
+   "data": {
+      "product": null
+   },
+   "errors": [
+      // details about failure in related_products field populated here
+   ]
+}
+```
+
+## Backwards Compatibility
+
+- It is a safe, non-breaking change to move a field from nullable to non-nullable
+- It is an unsafe, breaking change to move a field from non-nullable to nullable
+
+With this in mind, when in doubt, the safer route is to make a field nullable, with the agreement that we'll iterate and improve schemas as we get feedback on usage from clients.
+
+**Note**: The [GraphQL reference implementation](https://github.com/graphql/graphql-js) has a `findBreakingChanges` feature that can be inspected to see a [list of all known schema changes that cause breaks for clients](https://github.com/graphql/graphql-js/blob/2232ebdef828566f3add3ed2a31709d3c1710c0e/src/utilities/findBreakingChanges.js#L41-L67). It [explicitly notes](https://github.com/graphql/graphql-js/blob/2232ebdef828566f3add3ed2a31709d3c1710c0e/src/utilities/findBreakingChanges.js#L461) that moving from a nullable value to a non-nullable value of the same type is safe.
+
+## Recommendations
+
+### Top-level Query fields should always be nullable
```

Comment:
As a M2 GraphQL consumer I disagree. If a query is guaranteed to return data, it should not be null and making it nullable introduces unnecessary defensive code, slowing down development or worse risking null-errors due to excessive use of non-null-assertions.

Most of the time queries returning arrays will always return an array but empty when there are no results. e.g `countries: [Country!]!`. In this case I don't think there is any reason why `countries` should ever be null. There are probably others where this is the case.

I don't think there are a lot of queries (or mutations) in M2 that return a non-nullable singular type. I think `currency` and `storeConfig` are two that should never return `null`. I reckon the recommendation should be to look at this case-by-case. Anything that takes a param such as id to search for an item should be nullable (as the id may not exist and `null` is usually used for "not found" in GraphQL).